src/toolkit/res_buf.h

```diff
@@ -163,6 +172,37 @@ class ResBuf {
     return Pop(qty);
   }
 
+  /// Pops the specified amount of material, re-packed into the the 
+  /// package type provided. Returns any remaining material that doesn't
+  /// fit in a package per the packaging strategy and limits as a 
+  /// separate vector with default packaging
+  /// If it's not possible to fill the package
+  /// it returns an empty vector.
+  std::vector<typename T::Ptr> PopPackaged(double qty, Package::Ptr pkg) {
+    typename T::Ptr r = Pop(qty);
+    std::vector<typename T::Ptr> rs_pkgd;
+    typename T::Ptr r_pkgd;
+    
+    double fill_mass = GetFillMass(r, pkg);
+    if (fill_mass ==0) {
+      return rs_pkgd;
+    }
+
+    while (r->quantity() > pkg->fill_min()) {
+      double pkg_fill = std::min(r->quantity(), fill_mass);
+      r_pkgd = boost::dynamic_pointer_cast<T>(r->ExtractRes(pkg_fill));
+      r_pkgd->ChangePackageId(pkg->id());
+      rs_pkgd.push_back(r_pkgd);
+    }
+    // push leftover material back into the buffer
+    double remaining = r->quantity() - r_pkgd->quantity();
+    if (remaining > 0) {
+      rs_.push_front(boost::dynamic_pointer_cast<T>(r->ExtractRes(remaining)));
+    }
+
+    return rs_pkgd;
+  }
+
```

Comment:
```suggestion
class Package {
public:

  /**
   * @brief Repackages a given resource into the specified package.
   *
   * This function takes a resource and a package as input, and repackages the resource into the package. 
   * The function returns a vector of packaged resources. If there is any remaining quantity after repackaging, 
   * it is left in the original resource. If the resource is fully repackaged, the original resource is set to null.
   *
   * @param r A shared pointer to the resource to be repackaged.
   * @param pkg A shared pointer to the package into which the resource will be repackaged.
   * @return A vector of packaged resources.
   */

  std::vector<typename T::Ptr> repackage(typename T::Ptr r, Package::Ptr pkg) {
    std::vector<typename T::Ptr> rs_pkgd;
    typename T::Ptr r_pkgd;

    double fill_mass = GetFillMass(r, pkg);
    if (fill_mass == 0) {
      return rs_pkgd;
    }

    while (r->quantity() > pkg->fill_min()) {
      double pkg_fill = std::min(r->quantity(), fill_mass);
      r_pkgd = boost::dynamic_pointer_cast<T>(r->ExtractRes(pkg_fill));
      r_pkgd->ChangePackageId(pkg->id());
      rs_pkgd.push_back(r_pkgd);
    }
    // push leftover material back into the buffer
    double remaining = r->quantity() - r_pkgd->quantity();
    if (remaining > 0) {
      rs_.push_front(boost::dynamic_pointer_cast<T>(r->ExtractRes(remaining)));
    }

    return rs_pkgd;
  }
};


/**
 * @brief Pops a specified amount of material and repackages it.
 *
 * This function pops a specified amount of material from a container and repackages it into the provided package type.
 * The function returns a vector of packaged resources. If there is any remaining quantity after repackaging, it is left in the original resource.
 * If the resource is fully repackaged, the original resource is set to null.
 *
 * @param qty The quantity of material to be popped and repackaged.
 * @param pkg A shared pointer to the package into which the material will be repackaged.
 * @return A vector of packaged resources.
 */
std::vector<typename T::Ptr> PopPackaged(double qty, Package::Ptr pkg) {
  typename T::Ptr r = Pop(qty);
  
  // Use the repackage function to do the packaging
  std::vector<typename T::Ptr> rs_pkgd = pkg->repackage(r, pkg);
  
  return rs_pkgd;
}
```

Reply:
Also, the `Package` class already exists elsewhere in the project as part of @nuclearkatie 's PhD research

Reply:
And, sorry, @nuclearkatie. I truly respect your efforts and believe that you may have your own ideas in progress. I just added this suggestion because I've found that I've already done similar things before. nvm